libraries/state_history/include/eosio/state_history/log.hpp

```diff
@@ -62,57 +77,261 @@ struct state_history_log_header {
 static const int state_history_log_header_serial_size = sizeof(state_history_log_header::magic) +
                                                         sizeof(state_history_log_header::block_id) +
                                                         sizeof(state_history_log_header::payload_size);
-struct state_history_log_prune_config {
-   uint32_t                prune_blocks;                  //number of blocks to prune to when doing a prune
-   size_t                  prune_threshold = 4*1024*1024; //(approximately) how many bytes need to be added before a prune is performed
-   std::optional<size_t>   vacuum_on_close;               //when set, a vacuum is performed on dtor if log contains less than this many bytes
+
+namespace state_history {
+   struct prune_config {
+      uint32_t                prune_blocks;                  //number of blocks to prune to when doing a prune
+      size_t                  prune_threshold = 4*1024*1024; //(approximately) how many bytes need to be added before a prune is performed
+      std::optional<size_t>   vacuum_on_close;               //when set, a vacuum is performed on dtor if log contains less than this many bytes
+   };
+
+   struct partition_config {
+      fc::path retained_dir       = "retained";
+      fc::path archive_dir        = "archive";
+      uint32_t stride             = 1000000;
+      uint32_t max_retained_files = 10;
+   };
+} // namespace state_history
+
+using state_history_log_config = std::variant<std::monostate, state_history::prune_config, state_history::partition_config>;
+
+struct locked_decompress_stream {
+   std::unique_lock<std::mutex> lock; // state_history_log mutex
+   std::variant<std::vector<char>, std::unique_ptr<bio::filtering_istreambuf>> buf;
+
+   locked_decompress_stream() = delete;
+   locked_decompress_stream(locked_decompress_stream&&) = default;
+
+   explicit locked_decompress_stream(std::unique_lock<std::mutex> l)
+   : lock(std::move(l)) {};
+
+   template <typename StateHistoryLog>
+   void init(StateHistoryLog&& log, fc::cfile& stream, uint64_t compressed_size) {
+      buf = std::make_unique<bio::filtering_istreambuf>();
+      std::get<std::unique_ptr<bio::filtering_istreambuf>>(buf)->push(bio::zlib_decompressor());
+      std::get<std::unique_ptr<bio::filtering_istreambuf>>(buf)->push(bio::restrict(bio::file_source(stream.get_file_path().string()), stream.tellp(), compressed_size));
+   }
+
+   template <typename LogData>
+   void init(LogData&& log, fc::datastream<const char*>& stream, uint64_t compressed_size) {
+      buf = std::make_unique<bio::filtering_istreambuf>();
+      std::get<std::unique_ptr<bio::filtering_istreambuf>>(buf)->push(bio::zlib_decompressor());
+      std::get<std::unique_ptr<bio::filtering_istreambuf>>(buf)->push(bio::restrict(bio::file_source(log.filename), stream.pos() - log.data(), compressed_size));
+   }
```

Comment:
```suggestion
   template <typename StateHistoryLog>
   void init(StateHistoryLog&& log, fc::cfile& stream, uint64_t compressed_size) {
      auto istream = std::make_unique<bio::filtering_istreambuf>();
      istream->push(bio::zlib_decompressor());
      istream->push(bio::restrict(bio::file_source(stream.get_file_path().string()), stream.tellp(), compressed_size));
      buf = std::move(istream);
   }

   template <typename LogData>
   void init(LogData&& log, fc::datastream<const char*>& stream, uint64_t compressed_size) {
      auto istream = std::make_unique<bio::filtering_istreambuf>();
      istream->push(bio::zlib_decompressor());
      istream->push(bio::restrict(bio::file_source(log.filename), stream.pos() - log.data(), compressed_size));
      buf = std::move(istream);
   }
```

Just a suggestion, I think this is a little bit cleaner.

Reply:
Thanks!